src/main/webapp/src/app/admin-page/admin-page.component.ts

```diff
@@ -0,0 +1,341 @@
+// limitations under the License.
+
+import {Component, OnInit} from '@angular/core';
+import {ActivatedRoute} from '@angular/router';
+import {
+  HttpClient,
+  HttpResponse,
+  HttpErrorResponse,
+} from '@angular/common/http';
+import {Observable} from 'rxjs';
+import {Garden} from '../model/garden.model';
+import {Plant} from '../model/plant.model';
+import {User} from '../model/user.model';
+
+@Component({
+  selector: 'admin-page',
+  templateUrl: './admin-page.component.html',
+  styleUrls: [
+    './admin-page.component.css',
+    '../common/growpod-page-styles.css',
+  ],
+})
+
+/**
+ * This is a component designed to allow administrators to manage the gardens they
+ * own.
+ *
+ * This page takes an argument 'garden-id'. This is not optional, and determines
+ * the garden that is being administered. Only the garden owner can make changes to
+ * their garden.
+ */
+export class AdminPageComponent implements OnInit {
+  isLoaded = false;
+  gardenProfile: Garden | null;
+  gardenManager: string;
+
+  // User list
+  gardenUserIdList: Array<string> | null = null;
+  gardenUserNameMap: Map<string, string>;
+  gardenUserIdListError = 'User list not loaded';
+
+  // Plant list
+  gardenPlantIdList: Array<string> | null = null;
+  gardenPlantNameMap: Map<string, string>;
+  gardenPlantIdListError = '';
+  errorMessage = '';
+
+  /**
+   * Initializes the component based on provided arguments
+   *
+   * @param route Contains arguments.
+   */
+  constructor(private route: ActivatedRoute, private httpClient: HttpClient) {}
+
+  ngOnInit(): void {
+    const gardenIdArg = this.route.snapshot.paramMap.get('garden-id');
+    if (!gardenIdArg) {
+      this.gardenProfile = null;
+      this.errorMessage = 'No garden-id argument in the query string.';
+      return;
+    }
+    this.createAdminPage(gardenIdArg);
+  }
+
+  /**
+   * Gets garden information for the specified garden from
+   * the server. Returns an observable HTTP response.
+   *
+   * Performs GET: /garden/{garden}
+   *
+   * @param garden The garden reqested from the server.
+   * @return the http response.
+   */
+  getGardenInfo(garden: string): Observable<HttpResponse<Garden>> {
+    return this.httpClient.get<Garden>('/garden/' + garden, {
+      observe: 'response',
+      responseType: 'json',
+    });
+  }
+
+  /**
+   * Gets all users this garden contains.
+   *
+   * Performs GET: /garden/{garden}/user-list
+   *
+   * @return the http response.
+   */
+  getGardenUserList(garden: string): Observable<HttpResponse<Array<String>>> {
+    return this.httpClient.get<Array<String>>(
+      '/garden/' + garden + '/user-list',
+      {
+        observe: 'response',
+        responseType: 'json',
+      }
+    );
+  }
+
+  /**
+   * Gets all plants this garden contains.
+   *
+   * Performs GET: /garden/{garden}/plant-list
+   *
+   * @return the http response.
+   */
+  getGardenPlantList(garden: string): Observable<HttpResponse<Array<String>>> {
```

Comment:
note: this is interesting from a type perspective. I don't know the standard in terms of how to handle this in Angular, but I'd imagine best practice when creating an `Observable` here would be to also flatten the response. The ideal return type would be `Observable<String>`.

Why?

An observable represents a stream of values delivered asynchronously. In this instance they happen to all come at once via an `HttpResponse`, but you could imagine the stream of plant ids coming from anywhere. It would be better to deal with containing/unwrapping the HTTP bits here, so the client of this function only needs to deal with the asynchronously delivered strings.

---

If you don't know `rxjs` yet, I wouldn't recommend tightening up the types now, since there is a learning curve.

Reply:
I agree, your suggestion would have been better, and furthermore I don't think it would be *that* hard of a change to make (??) But it would simply take up too much engineering time for the last week.